httpx/http_service.go

```diff
@@ -0,0 +1,208 @@
+type ResponseError struct {
+	response *http.Response
+	err      error
+}
+
+// A RoundTrip implementation that can handle both normal responses and cancellations.
+func (t *Transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
+	var returnResp *http.Response
+	var returnErr error
+
+	// Run the HTTP request in a goroutine and pass the response by a channel.
+	tr := t.Client.Transport.(interface {
+		CancelRequest(*http.Request)
+	})
+	c := make(chan ResponseError, 1)
+
+	go func() {
+		resp, err := t.Client.Do(req)
+		re := ResponseError{response: resp, err: err}
+		c <- re
+	}()
+
+	select {
+	case <-ctx.Done():
+		tr.CancelRequest(req)
+		<-c
+		returnResp = nil
+		returnErr = ctx.Err()
+	case respError := <-c:
+		returnResp = respError.response
+		returnErr = respError.err
+	}
+	return returnResp, returnErr
+}
+
+// RetryTransport is an implementation of the RoundTripper interface that
+// retries requests.
+type RetryTransport struct {
+	*retry.Retrier
+	Transport RoundTripper
+}
+
+func (t *RetryTransport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
+	resp, err := t.Retrier.Retry(func() (interface{}, error) {
+		resp, err := t.Transport.RoundTrip(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode >= 500 {
+			return nil, &RetryableHTTPError{Path: req.URL.String(), StatusCode: resp.StatusCode}
+		} else if resp.StatusCode >= 300 {
+			return nil, &HTTPError{Path: req.URL.String(), StatusCode: resp.StatusCode}
+		}
+
+		return resp, nil
+	})
+
+	if resp == nil {
+		return nil, err
+	} else if response, ok := resp.(*http.Response); ok {
+		return response, err
+	} else {
+		panic("Response is non-nil and not of an expected type")
+	}
+}
+
+// RequestIDTransport is an http.RoundTripper implementation that adds the
+// embedded request id to outgoing http requests.
+type RequestIDTransport struct {
+	Transport RoundTripper
+}
+
+func (t *RequestIDTransport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
+	req.Header.Add("X-Request-Id", RequestID(ctx))
+	return t.Transport.RoundTrip(ctx, req)
+}
+
+// NewJSONRequest generates a new http.Request with the body set to the json
+// encoding of v.
+func NewJSONRequest(method, path string, v interface{}) (*http.Request, error) {
+	var r io.Reader
+	if v != nil {
+		raw, err := json.Marshal(v)
+		if err != nil {
+			return nil, err
+		}
+		r = bytes.NewReader(raw)
+	}
+
+	req, err := http.NewRequest(method, path, r)
+	if err != nil {
+		return nil, err
+	}
+	if v != nil {
+		req.Header.Set("Content-Type", "application/json")
+	}
+	return req, nil
+}
+
+func ParseURL(baseURL *url.URL, path string) string {
+	urlString := UrlWithoutCreds(*baseURL)
+	return strings.TrimRight(urlString, "/") + path
+}
+
+// Shows the URL without information about the current username and password
+func UrlWithoutCreds(u url.URL) string {
```

Comment:
Minor style nitpick. In Go, it's preferred to use URL instead of Url (and ID instead of Id, etc).